processing/src/main/java/io/druid/query/groupby/orderby/DefaultLimitSpec.java

```diff
@@ -149,35 +154,100 @@ public int compare(Row left, Row right)
       @Override
-      public int compare(Row left, Row right)
+      public Ordering<Row> apply(OrderByColumnSpec input)
       {
-        return comparator.compare(left.getRaw(column), right.getRaw(column));
+
+        return new Ordering<Row>()
+        {
+          @SuppressWarnings("unchecked")
+          @Override
+          public int compare(Row left, Row right)
+          {
+            return comparator.compare(left.getRaw(column), right.getRaw(column));
+          }
+        };
       }
     };
   }
 
-  private Ordering<Row> dimensionOrdering(final String dimension)
+  private Function<OrderByColumnSpec, Ordering<Row>> dimensionOrdering(final String dimension)
   {
-    return Ordering.natural()
-                   .nullsFirst()
-                   .onResultOf(
-                       new Function<Row, String>()
-                       {
-                         @Override
-                         public String apply(Row input)
-                         {
-                           // Multi-value dimensions have all been flattened at this point;
-                           final List<String> dimList = input.getDimension(dimension);
-                           return dimList.isEmpty() ? null : dimList.get(0);
-                         }
-                       }
-                   );
+    return new Function<OrderByColumnSpec, Ordering<Row>>()
+    {
+      @Override
+      public Ordering<Row> apply(OrderByColumnSpec columnSpec)
+      {
+        if (columnSpec.asNumber()) {
+          // Order: null < string < number , and number only means "long" here.
+          return new Ordering<Comparable>()
+          {
+            @Override
+            public int compare(
+                Comparable left, Comparable right
+            )
+            {
+              if (left instanceof Long && right instanceof Long) {
+                return ((Long) left).compareTo((Long) right);
+              } else if (left instanceof String) {
+                return -1;
+              } else if (right instanceof String) {
+                return 1;
+              } else {
+                return ((String) left).compareTo((String) right);
+              }
+            }
+          }.nullsFirst().onResultOf(
+              new Function<Row, Comparable>()
+              {
+                @Override
+                public Comparable apply(Row input)
+                {
+                  // Multi-value dimensions have all been flattened at this point;
+                  final List<String> dimList = input.getDimension(dimension.toLowerCase());
+                  if (dimList.isEmpty()) {
+                    return null;
+                  }
+
+                  String dimValue = dimList.get(0);
+                  // Make sure dimValue is number parsable.
+                  if (dimValue.isEmpty()) {
+                    return dimValue;
+                  }
+                  for (int i = 0; i < dimValue.length(); i++) {
```

Comment:
It does not seem to hard to make this feature complete by adding support for doubles as well.
We can use a regex here instead to detect if its a double value or long value or something else. Accordingly use Double.parseDouble(..) or Long.parseLong(..) .
And in the compare above add another case to handle Doubles.
Also, you can refactor this code a little bit by having a private method like
private Comparable getDimensionValue() {

}
and calling return getDimensionValue(dimList.get(0));